AI Chat + AI Tutor: Add multimodal gemini support

edcodedotorg opened the pull request:
This ~is a WIP~ PR ~for discussion which~ adds Google Gemini support to AI Chat and replaces [this PR](https://github.com/code-dot-org/code-dot-org/pull/66254) by adding multimodal support which also required using the actual Gemini API endpoint:

 e.g. `https://generativelanguage.googleapis.com/v1beta/models/...`

vs the Open ID Compatibility API endpoint:

 `https://generativelanguage.googleapis.com/v1beta/openai/chat/completions`

https://github.com/user-attachments/assets/4086552a-82cb-41ff-bf1c-8151f6d9a753

This PR also represents a significant refactoring in order to share code between OpenAI and Google endpoints so I'm offering it for comments as I work through next steps.  

Next steps are to complete various TODOs within the code ~~as well as:~~

1. ~~Make Gemini the default for AI Tutor and~~
2. ~~Add query param option to switch back to OpenAI or LearnLM.~~
[edit: these were added [in this commit](https://github.com/code-dot-org/code-dot-org/pull/66675/commits/672ade83b7379ccfd89d5500d36aa6eabb3ed7ed)]

Note: this PR leaves multiple uses of the old `AichatOpenaiResponsesHelper` e.g. in `aichat_safety_helper.rb` and `openai_pdf_processesing_availability_helper.rb` and also does not add SageMaker models to the new combined class which can both be dealt with in a later work.

To test, you will need to update/create an AI Chat level w/ multimodal  support enabled along with one or more google models.  Remember that your level should only have multimodal enabled models for multimodal support to work.  You can use this [file](https://raw.githubusercontent.com/code-dot-org/code-dot-org/187df2b6e9be052a9abe155661d02e61d75dd298/dashboard/config/levels/custom/aichat/AI%20Chat%20Test%20Level.level) and replace `dashboard/config/levels/custom/aichat/AI Chat Test Level.level` to make it easier.

sanchitmalhotra126 commented on dashboard/app/helpers/aichat_ai_client.rb:
```diff
@@ -0,0 +1,270 @@
+# Combined class to hold shared logic for multiple AI API backends. This is essentially an "abstract" class
+# that is never instantiated directly. The derived classes hold implementation details in required overridden
+# methods. Currently the two implemented APIs (OpenAI and Gemini) are POST based REST APIs
+class AichatAiClient
```
If we're doing this refactor, we might as well go all the way and roll in the sagemaker clients into to this too? They also use REST APIs to call the sagemaker endpoints

edcodedotorg replied:
Yes, this is what I was referring to here:

> Note: this PR leaves multiple uses of the old AichatOpenaiResponsesHelper e.g. in aichat_safety_helper.rb and openai_pdf_processesing_availability_helper.rb and also does not add SageMaker models to the new combined class which can both be dealt with in a later work.

I talked to Brendan and he suggested I create a branch off of this branch to do the Sagemaker stuff and if it is ready in time it can be folded into this PR.  But please take a look at `aichat_safety_helper.rb` and `openai_pdf_processesing_availability_helper.rb` to discuss if/when those should be considered to be folded into this.

edcodedotorg replied:
added to TODO comment in PR.  resolving this one

sanchitmalhotra126 commented on dashboard/app/helpers/aichat_asset_helper.rb:
```diff
@@ -1,33 +1,104 @@
+   )
+    end
+
+    # NOTE: this can also result in other various errors,
+    # TODO - we should probbaly catch them and throw a unified error so that we can get a response back to the user
```
Note that in many cases, errors are deliberately left uncaught since they will eventually be handled by the `rescue_from` block in the AI Chat active job class, and logged and reported to our error reporting platform (HoneyBadger). Additionally, we do produce a generic error message for the user (keyed off of `AI_REQUEST_EXECUTION_STATUS[:FAILURE]`). While I agree that in some cases, specific error types are useful for testing and/or logging, in a lot of these cases, these errors ultimately represent some sort of internal system error where the user can't meaningfully do anything differently to fix it. In these cases, I think it's useful to just let the original error bubble up so we can get the relevant details and stack trace in HoneyBadger, as the user-facing error message will be the same regardless.

edcodedotorg replied:
> these errors ultimately represent some sort of internal system error where the user can't meaningfully do anything differently to fix it.

sure, but in this case we know that something is failing with fetching a file that the user "attached" to a chat message.  As a user, if the error message pointed to that and I kept getting an error I could try a different file or try without any file.  Error messages empower users to find workarounds on the spot. Another reason to be as descriptive as possible with error messages is that it helps us become aware of issues as they come through support channels which could be the reason we end up looking into the HoneyBadger logs.  

Nothing we have to add now, but I think it is something we should tighten up eventually.

sanchitmalhotra126 replied:
Yep, makes sense - as discussed, in this case I can definitely see a case for adding a new user-specific message to indicate that asset uploading in general might be broken or that they should try another file. Just worth noting that if we get to this state, it does represent some sort of internal or downstream error as we expect the file was uploaded successfully and present in the student's project asset bucket.

edcodedotorg replied:
thanks.  Just adding a few asset related observations here for posterity:

1) (we discussed this) the way we are storing assets it seems that if the user ever uploads 2 or more files with the same name as part of a chat, the latest one will be passed to the third party endpoint multiple times and the previous one(s) will be lost

2) (we also discussed this) the way we are storing assets and passing them to the third party backend means we will send the same file multiple times in a given chat (as part of the history) and potentially means the file will be stored with the project indefinitely

3) (forgot to mention this) from a quick reading of the code it seemed that a message can time out or otherwise fail but still be part of the history.  I may be wrong but that's the impression that I got when following the code (I'll look again now that I understand it better)

4) (forgot to mention this) during testing I noticed a frontend glitch that caused the file that was previously added to disappear from the chat before sending the message.  I'll try to replicate and add a video

edcodedotorg replied:
adding observations to TODO PR comment

sanchitmalhotra126 commented on dashboard/app/helpers/aichat_ai_client.rb:
```diff
@@ -0,0 +1,270 @@
+class AichatAiClient
+  # get an instance of the appropriate derived class based on model id
+  def self.get_instance(model_id)
+    # TODO - we should create a map/hash from model_id to model/key (not sure why model id isn't just model but whatever)
```
Curious what `model/key` refers to here? The provided `model_id` should in theory be one of the constants defined in `AI_CHAT_MODEL_IDS`.

edcodedotorg replied:
Sorry, `model/key` here refers to the actual model and key sent to the third party AI API (OpenAI, Gemini)

Specifically I'm referring to the fact that for OpenAI, the `model_id` we pass from the frontend  is `'gpt-4o-mini'` but we are sending a [specific model version](https://github.com/code-dot-org/code-dot-org/blob/b01c2f9d330d3098752da56ad171e0fecee330e5/lib/cdo/shared_constants.rb#L738) to the API  which is hard coded in `AICHAT_MODEL_VERSION` (i.e. `'gpt-4o-mini-2024-07-18'`).  

My _question_ is: why is `'gpt-4o-mini'`  our internal "model ID" where `'gpt-4o-mini-2024-07-18'` is the real OpenAI model ID.  What's the point of hardcoding `AICHAT_MODEL_VERSION = 'gpt-4o-mini-2024-07-18'` in `shared_constants.rb` just to have to look it up when we see a model ID of `'gpt-4o-mini'`?  

If we want this level of indirection, I would take my _suggestion_ of `a map/hash from model_id to model/key` a bit further to  something like `a map/hash from model_id to model/key/provider`  (without this model indirection `a map/hash from model_id to key/provider` would still be handy) which would enable:

```
if model_id == "gpt-4o-mini"
  return AichatOpenaiCompletionsClient.new(CDO.openai_student_learning_api_key, SharedConstants::AICHAT_MODEL_VERSION)
else
  return AichatGeminiClient.new(CDO.google_gemini_student_learning_api_key, model_id)
end
```

to become:

```
model_details = model_key_provider[model_id]
if model_details.provider == SharedConstants::OPENAI
  return AichatOpenaiCompletionsClient.new(model_details.key, model_details.model)
else
  return AichatGeminiClient.new(model_details.key, model_details.model)
end
```

Alternatively, if we ever want to allow a given model to be accessible for multiple third party backends (e.g. for open source models), we could just pass model and provider in directly and only look up the key.

sanchitmalhotra126 replied:
Discussed offline - decoupling the model ID from the model version allows us to switch the model version under the hood if we want, since we don't expect to actually expose different specific model versions to levelbuilders and/or students. Agreed with your suggestion of using a hash though! Will be especially useful as we also need to consider the "project" or "product" type to look up the right key/version.

sanchitmalhotra126 replied:
nit: mind cleaning up/updating this TODO to reflect the intended plan?

sanchitmalhotra126 commented on dashboard/app/helpers/aichat_ai_client.rb:
```diff
@@ -0,0 +1,270 @@
+    # TODO - we should create a map/hash from model_id to model/key (not sure why model id isn't just model but whatever)
+    # but for now we just assume it's one of the gemini models if not 'gpt-4o-mini'
+    if model_id == "gpt-4o-mini"
+      return AichatOpenaiCompletionsClient.new(CDO.openai_student_learning_api_key, SharedConstants::AICHAT_MODEL_VERSION)
```
`AICHAT_MODEL_VERSION` seems like a good candidate for a rename too (at some later point) as it's Open AI specific 🙂

edcodedotorg replied:
yeah this would get replaced with the hash mentioned above.

sanchitmalhotra126 commented on dashboard/app/helpers/aichat_ai_client.rb:
```diff
@@ -0,0 +1,270 @@
+      url,
+      headers: headers,
+      body: body.to_json,
+      open_timeout: DCDO.get('openai_http_open_timeout', 5),
+      read_timeout: DCDO.get('openai_http_read_timeout', 30)
```
Should these be overridden/set by subclasses?

edcodedotorg replied:
Sorry I should have added a comment there.  I don't know anything about `DCDO.get` so didn't want to touch it

sanchitmalhotra126 replied:
Gotcha - DCDO is our configuration management system, where we can read and set simple production values without doing a full deploy. Follows the format `stored_value = DCDO.get(<key name>, <default value>)`. It's probably fine to leave these as is for now, but we might want to at least tweak the key name so it's not OpenAI specific.

edcodedotorg replied:
Ok thanks, I wasn't sure if tweaking the names would break anything.  But sure let's do it.  Do we want to give the shared AI Chat / Tutor plumbing a name and use that as the prefix?  Or is `ai_http_open_timeout` sufficient?

sanchitmalhotra126 replied:
`ai_http_open/read_timeout` seems good for now.

edcodedotorg replied:
will look at this in the morning

edcodedotorg replied:
@sanchitmalhotra126 hey I still don't understand this enough to change it confidently.  I don't see any place this is `set` only these `gets`.  Is the effect of `{ open_timeout: DCDO.get('openai_http_open_timeout', 5) }` the same as just saying `{ open_timeout:  5 }`?

<img width="484" alt="Screenshot 2025-06-27 at 10 03 11 AM" src="https://github.com/user-attachments/assets/a1584fdb-a6c7-4376-8d12-8d9f3a4f7575" />

sanchitmalhotra126 replied:
No worries, we can leave as-is for now and file a TODO to follow-up later, as I imagine this may be slightly tweaked anyway because SageMaker will not use the HTTP client.

These values are optionally `set` in the admin console in our various prod and non-prod environments (I can show you how to access that offline if you're interested). If no value is set, we use the default, which is the second argument to the `get` function.

edcodedotorg replied:
added TODO to PR comment and closing this out here

sanchitmalhotra126 commented on dashboard/app/helpers/aichat_ai_client.rb:
```diff
@@ -0,0 +1,270 @@
+    # Prompt tokens are by far and away our largest cost driver (and the piece that users actually control),
+    # so we throttle on that.
+    limit = DCDO.get('aichat_token_limit_per_day', DEFAULT_TOKEN_LIMIT_PER_DAY)
+    Cdo::Throttle.throttle(AichatOpenaiHelper.token_throttling_key(model_id, user_id),
```
Maybe for later - we should move `token_throttling_key` out of `AichatOpenaiHelper` as it's not OpenAI-specific

edcodedotorg replied:
adding to general TODO in PR - would be solved along with removing helper altogether

sanchitmalhotra126 commented on dashboard/app/helpers/aichat_ai_client.rb:
```diff
@@ -0,0 +1,270 @@
+    #
+    # Prompt tokens are by far and away our largest cost driver (and the piece that users actually control),
+    # so we throttle on that.
+    limit = DCDO.get('aichat_token_limit_per_day', DEFAULT_TOKEN_LIMIT_PER_DAY)
+    Cdo::Throttle.throttle(AichatOpenaiHelper.token_throttling_key(model_id, user_id),
```
Could we perform the throttling in a different function now that we're calling into the subclass to compute the usage data for us?

edcodedotorg replied:
sorry trying to go through these rapid fire, and am not sure I understand.  Probably obvious but could you explain a little more?

sanchitmalhotra126 commented on dashboard/app/helpers/aichat_ai_client.rb:
```diff
@@ -0,0 +1,270 @@
+    # *responses* even if we have a way to disable them in level builder
+    response_text = extract_text_response_from_body(response_body)
+
+    # TODO - add back report_usage_and_throttling_metrics
+    # This requires some thought around how to make it implementation agnostics re: 'messages' and 'usage'
```
Is this TODO still relevant?

edcodedotorg replied:
nope - removed

sanchitmalhotra126 commented on dashboard/app/helpers/aichat_openai_helper.rb:
```diff
@@ -3,190 +3,56 @@
-  MODEL = SharedConstants::AICHAT_MODEL_VERSION
+# from AI Chat lab to be sent to the AI API, and then sends the request to the API.
 
+module AichatOpenaiHelper
```
Could we kill this helper entirely? Seems like most of what it does (unwrapping `aichat_customizations`) could be done in the new client or the active job code.

edcodedotorg replied:
yeah I was thinking the same thing ;). Either push up to active job or down into the shared class (or combination of the two).

Since the active job code switches between this and Sagemaker, I can look at doing so as part of the Sagemaker refactor.  Currently this helper and the Sagemaker helper function are pretty close in shape so I wanted to leave it as-is until I tackled that.

edcodedotorg replied:
moved this to general TODO Comment

sanchitmalhotra126 commented on dashboard/app/helpers/aichat_ai_client.rb:
```diff
@@ -0,0 +1,270 @@
+
+    raise_possible_response_errors_from_body(response_body)
+
+    # TODO - the fact we called this extract_text_response_from_body suggests we need to adapt this class for multimodal
+    # *responses* even if we have a way to disable them in level builder
```
We don't currently have any support for multimodal/non-text responses (only inputs)

edcodedotorg replied:
sure.. I guess I'm just just thinking out loud about things we know are possible with the 3rd party APIs that we can start to think about exposing even if just for exploration.  You mentioned that one problem is that it's hard to know where we are going next but one way to look at this backend code is that it's just here to expose the things the 3rd party APIs expose, but in a unified way.  This could be "for future use" but it helps to keep all the possibilities in mind.

sanchitmalhotra126 replied:
Yep sure! Agreed that we'd likely want to update the function name if/when we support multimodal responses, not to mention all the other work to support passing that binary data back to the user 🙂

sanchitmalhotra126 replied:
nit: mind removing/updating this TODO? Assuming this is referring to multimodal responses, we don't currently have "a way to disable them in levelbuilder" (this could instead be something like "update the function name when supporting multimodal responses" or similar?)

sanchitmalhotra126 commented on dashboard/app/helpers/aichat_ai_client.rb:
```diff
@@ -0,0 +1,270 @@
+# methods. Currently the two implemented APIs (OpenAI and Gemini) are POST based REST APIs
+class AichatAiClient
+  # get an instance of the appropriate derived class based on model id
+  def self.get_instance(model_id)
```
nit: this isn't really a singleton as we're instantiating a new client on every call. Maybe `create` or similar?

edcodedotorg replied:
sure yeah it's the factory pattern.  And no problem..  haha I must have forgot `get_instance` was so common with the singleton pattern that it actually _signals_ that it's a singleton.  You'll notice with my naming that I try to be as descriptive as possible and `create` doesn't tell me what I'm creating.  Any opposition to `create_instance`?

BTW: I use this pattern to lazy load singletons in Typescript in a way that only the class name could tell you that it's a singleton so I mostly include `Singleton` (or sometimes `Manager`) in the name (not sure if this is common or not):

```
class TestSingleton {

    static #instance?: TestSingleton;

    constructor(){
        if(!TestSingleton.#instance){
            TestSingleton.#instance = new TestSingleton;
        }
        return TestSingleton.#instance;
    }
}

const testSingleton = new TestSingleton;
```

sanchitmalhotra126 replied:
Sure thing, `create_instance` is fine - my main callout is just that this function creates a fresh instance each time its invoked, as opposed to the typical singleton pattern of returning the singleton instance once created (like your TypeScript example).

edcodedotorg replied:
ok sorry I thought your callout was that I should rename it since it wasn't a singleton, but did you think we should use a singleton?  I just matched the existing code that creates an instance per invocation and didn't give it much thought.

sanchitmalhotra126 replied:
Oh haha no no, my callout was purely naming 🙂 - the implementation and new name you have is fine with me!

edcodedotorg replied:
updated to `create_instance` - hiding this comment as resolved

sanchitmalhotra126 commented on dashboard/config/levels/custom/aichat/AI Chat Test Level.level:
```diff
@@ -57,4 +57,4 @@
   "audit_log": "[{\"changed_at\":\"2023-04-05 09:03:03 +0000\",\"changed\":[\"published\"],\"changed_by_id\":4,\"changed_by_email\":\"brendan+levelbuilder@code.org\"},{\"changed_at\":\"2023-08-03 15:36:40 -0700\",\"changed\":[],\"changed_by_id\":1,\"changed_by_email\":\"sanchit+localhost@code.org\"},{\"changed_at\":\"2023-08-14 15:34:06 -0700\",\"changed\":[],\"changed_by_id\":1,\"changed_by_emai ...
 }]]></config>
-</Aichat>
+</Aichat>
```
nit: mind reverting so this file isn't in the diff 🙂

edcodedotorg replied:
yeah I don't think I can technically `revert` it as that wasn't the only thing in the commit that initially changed it.  I don't actually see the difference in the diff above but I'll double check so that once squashed it doesn't show up

edcodedotorg replied:
ok looks like there was a newline at end of the file in one of the versions.  I updated and it no longer shows up in changed files.  Closing this comment as resolved

sanchitmalhotra126 commented on dashboard/app/helpers/aichat_ai_client.rb:
```diff
@@ -0,0 +1,270 @@
+    }
+  end
+
+  # TODO - rework this to make it provider agnostic and add call
```
This definitely seems like something we should do as part of adding Gemini support as we will be reporting metrics as soon as we start using it.

edcodedotorg replied:
oh sorry, that's mostly a TODO that I already finished (it's now set up to report for Gemini).  But I left it there to remember to discuss how we want to differentiate the logs.  

Currently it logs `OpenAI response...` to Honeybadger and uses  `metric_name: "AichatOpenaiRequest.#{key}"`.  

For honeybadger I can switch the log text between provider name (OpenAI/Gemini) but I'm assuming we want  a metric per provider?  Should we keep it `AichatOpenaiRequest` for OpenAI and `AichatGeminiRequest` for Gemini?  What about per product (`AiTutorGeminRequest`).  

Also is PascalCase typical for how we write the metric names or are dots separating the parts more common? (e.g. `"airequest.gemini.aitutor.#{key}"`?

sanchitmalhotra126 replied:
Yeah looks like there are a few things we need to genericize:
- currently uses/assumes OpenAI GPT4o-mini token costs
- log payload event name is `aichat_openai_usage` - probably want to change to something more general `aichat_usage` with an additional field indicating the model
- DCDO flag is called `log_aichat_openai_usage`
- The metric name as you called it is also OpenAI-specific; i'd suggest we make a generic name (`AichatRequest.#{key}`) and add model as a new dimension.

Note that the above will also require us to update our queries and dashboards. One option would be to temporarily skip this reporting for Gemini and leave the code as-is, and follow-up with updates to make the reporting provider-agnostic.

sanchitmalhotra126 replied:
AFAIK, PascalCase with dots is pretty typical for cloudwatch metrics, at least from what I've seen?

edcodedotorg replied:
@sanchitmalhotra126 just pushed a commit to disable the throttling/metrics call for gemini temporarily as discussed.  Can you give me a quick run down of what it would take to update the queries and dashboards and I can re-enable it in a further PR.

bencodeorg commented:
Looks like the change to modelDescriptions.json makes these models available to levelbuilders, which I don't think we want yet? When we first added OpenAI we added a giant disclaimer like [LEVELBUILDER USE ONLY...] or something to the model name so they didn't end up in real levels (yet).

![image](https://github.com/user-attachments/assets/e22756e2-4f2c-40f8-a561-0919f1ba2678)

bencodeorg commented on dashboard/app/helpers/aichat_openai_helper.rb:
```diff
@@ -3,190 +3,56 @@
-    end
-    formatted
-  end
+    client = AichatAiClient.get_instance(model_id)
```
Am I reading this right that we get the Google client within `AichatOpenaiHelper`?

edcodedotorg replied:
yes there were open questions about naming/folder location. This is also plumbing for AI Tutor not just Chat so the name hasn't been considered yet at all.  What do you propose that covers:

1) describes AI Chat and AI Tutor
2) describes Google/OpenAI/Sagemaker/etc

edcodedotorg replied:
also as you can see, sometimes during refactoring, files don't last long enough for the name to matter: https://github.com/code-dot-org/code-dot-org/pull/66675#discussion_r2164674255

But in this case figuring out answers for 1+2 above are needed for the other files

bencodeorg replied:
Agree that removing this helper would solve the problem (and doesn't seem wildly out of scope given the breadth of changes here)? I guess I'm a little concerned about leaving a series of refactors mostly done if, say, next week we're all assigned to work on the project Karim emailed about yesterday :)

edcodedotorg replied:
Understood... I'll find the time to get this all to a good place.  I spent a bunch of time on this last weekend, and am just chalking it up to learning ruby and the codebase.

bencodeorg replied:
(please don't feel pressure to work on this on weekends unless you want to!)

edcodedotorg replied:
of course, just saying I find this refactor important and am willing to find the time.  Just renamed to AichatAiClient since it didn't make the chopping block this PR.

edcodedotorg commented on dashboard/test/helpers/aichat_openai_helper_test.rb:
@sanchitmalhotra126 hey I removed this file because 

1) the file it tests is planned to go away and 
2) it only tests formatting of messages which is now within a private method (where private methods are typically not tested).

Test coverage in this backend code already wasn't too great and now this means we have less tests (even though the three tests in this file were testing what are ultimately internal/private concerns)

So how should we go about getting a solid baseline of coverage?  There is only one public method in `aichat_ai_client` and its derived classes: `get_response_text`, which hits the network (to OpenAI / Google) so would likely have to be stubbed.  There doesn't seem to be any existing unit/integration test that stubs out this OpenAI call (there is a `StubbedSagemakerClient` for UI tests though but nothing that I can see that tests that e.g. the request body is properly formed). 

So my basic approach would be to stub the network calls out for each 3rd party AI API endpoint to ensure that we're generating request bodies that have the expected shape and that a non-error response would get back to the caller where a response with an error object in the response body will raise appropriately (A side note is the [existing code](https://github.com/code-dot-org/code-dot-org/blob/23d5fdf483e821c2001838f4c6e023f1c61c2cbd/dashboard/app/helpers/aichat_openai_helper.rb#L80) never checked the `http_response.code` and went straight to trying to parse the body as JSON so if the response is not JSON I guess this is the "let it bubble to Honeybadger" approach - as discussed I mostly left things like this as is for now.)

What are your thoughts/suggestions on this approach to stubbing? It seems some code in the codebase uses `WebMock` so I could put something together with that. Now that we have slightly less coverage than before, my thought is I should prioritize these tests over the Sagemaker stuff but let me know.

edcodedotorg replied:
This [commit](https://github.com/code-dot-org/code-dot-org/pull/66675/commits/41867a6759566ceedcd429c2c2775818c8515bf6) adds a first pass in this direction (checks for correct request body in stubbed network call and given the stubbed response, it checks the correct return value from `get_response_text`, but doesn't yet test the error case).  Let me know what you think

sanchitmalhotra126 replied:
This seems fine; I think modularizing the "AI Client" code further as we pull in sagemaker and create more subclasses will allow us to set up better test coverage as well.

edcodedotorg replied:
@sanchitmalhotra126 the l[atest commit](https://github.com/code-dot-org/code-dot-org/pull/66675/commits/001f14dc33dbf6e991fdd3ec6d80d2668f1d0ec6) adds gemini and the error testing

edcodedotorg commented:
> Looks like the change to modelDescriptions.json makes these models available to levelbuilders, which I don't think we want yet? When we first added OpenAI we added a giant disclaimer like [LEVELBUILDER USE ONLY...] or something to the model name so they didn't end up in real levels (yet).
> 

hey I'm honestly a little unsure who levelbuilders are if not code.org employees.  Are external people / teachers able to build their own levels and courses?  

But either way I take it bad things can/have happen(ed) if we expose something to levelbuilders that wasn't meant for prime time, so whatever you think makes sense.  I removed the level itself I used for testing from the PR after talking to Sanchit so we could do the same thing with `modelDescriptions.json` if you and Bryan aren't ready to test the gemini support in AI Chat more widely.  IIRC that file isn't needed for the existing chat plumbing we use for AI Tutor but is required to test in AI Chat.

If you want to leave it in an add a message, perhaps `[DEVELOPMENT USE ONLY...]` or `[DO NOT USE...]` rather than  `[LEVELBUILDER USE ONLY...]` as I thought your point was level builders should not be using these 🤷🏽‍♂️

edcodedotorg commented on apps/src/levelbuilder/ai-iteration-tools/ai-tutor/constants.ts:
```diff
@@ -36,6 +36,14 @@ const genAIEndpoints = [
     id: AiChatModelIds.CHATGPT,
     name: 'ChatGPT + Webpurify',
   },
+  {
```
@bencodeorg your comment about modelDescription.json reminded me that I was wondering about this file and the `+ Webpurify` pattern here. Upon review, I don't think this file is even related to the AI Chat app or plumbing, is it?

bencodeorg replied:
I don't really know what this is 😬 , agree it's not related to AI Chat. I think maybe something @Erin007 put together to help levelbuilders (aka curriculum) at one point?

bencodeorg replied:
haha recalling that I had same question when I added OpenAI:

https://github.com/code-dot-org/code-dot-org/pull/63565#discussion_r1934329184

Here's a DM where I asked her about it:

<img width="1271" alt="image" src="https://github.com/user-attachments/assets/b5e06ec1-bfd8-41aa-a8c3-446d2a507df8" />

And the PR she linked, which is now merged: https://github.com/code-dot-org/code-dot-org/pull/63368

edcodedotorg replied:
ok thanks!  So I'll "revert" that change too

Erin007 replied:
If ya'll aren't using the AI Tutor Tester code, you can go ahead an rip it out if it makes refactoring easier as far as I'm concerned.

edcodedotorg replied:
thanks Erin! We're basically not touching anything on AI Tutor 1.0 at the moment so I should be able to leave this file as it was.  Will confirm soon

edcodedotorg replied:
ok, now I remember why I added that, we can't build `/apps` without it as the model ids are used in a type assertion for the ids of the objects in that `genAIEndpoints` array, so that assertion is expecting the ids to match:

<img width="918" alt="Screenshot 2025-06-26 at 9 07 26 AM" src="https://github.com/user-attachments/assets/730edb68-7540-4221-ab67-b69bd32112ff" />

If leaving the gemini models in `genAIEndpoints` isn't a problem for legacy AI Tutor, I'd rather leave it as legacy AI Tutor will go away completely pretty soon.

bencodeorg replied:
Yeah that's what I did when I added OpenAI (added model to `genAIEndpoints` to satisfy TypeScript).

edcodedotorg replied:
ok thanks! closing this as resolved

edcodedotorg commented on apps/static/aichat/modelDescriptions.json:
@bencodeorg @dju90 this now has the `EXPERIMENTAL`  warning text.  Also, please note I copied the structure from openAI which has `...` for `overview` and `trainingData`. It seems `trainingData` wouldn't apply but `overview` could.  

Is there a specific use case for these fields that aren't generally applicable? If so, this is probably a sign we should separate that out somehow.

```
 {
    "id": "gpt-4o-mini",
    "name": "GPT-4o mini",
    "overview": "...",
    "trainingData": "...",
    "multimodal": true
  },
  {
    "id": "learnlm-2.0-flash-experimental",
    "name": "[EXPERIMENTAL, LEVELBUILDER USE ONLY] LearnLM 2.0 Flash",
    "overview": "...",
    "trainingData": "...",
    "multimodal": true
  },
  {
    "id": "gemini-2.0-flash",
    "name": "[EXPERIMENTAL, LEVELBUILDER USE ONLY] Gemini 2.0 Flash",
    "overview": "...",
    "trainingData": "...",
    "multimodal": true
  },
  {
    "id": "gemini-2.5-flash",
    "name": "[EXPERIMENTAL, LEVELBUILDER USE ONLY] Gemini 2.5 Flash",
    "overview": "...",
    "trainingData": "...",
    "multimodal": true
  },
  {
    "id": "gemini-2.5-pro",
    "name": "[EXPERIMENTAL, LEVELBUILDER USE ONLY] Gemini 2.5 Pro",
    "overview": "...",
    "trainingData": "...",
    "multimodal": true
  }
```

edcodedotorg replied:
leaving this open question about  `overview` and `trainingData` but closed other levelbuilder comments as resolved

bencodeorg replied:
> _Is there a specific use case for these fields that aren't generally applicable? If so, this is probably a sign we should separate that out somehow._

Yeah they are used in the "Compare Models" dialog if you are on a level ([example](https://studio.code.org/courses/allthethingscourse/units/1/lessons/47/levels/1)) where multiple models are available (you only choose between models in our Exploring Generative AI curriculum, as far as I know).

<img width="855" alt="image" src="https://github.com/user-attachments/assets/ca2b8b95-a8e2-4d49-b522-df104ea62238" />

Agree this file should probably be changed / is too broadly used (eg, causing the TypeScript errors with the AI Tutor Tester that we were chatting about with Erin)

edcodedotorg replied:
@bencodeorg so does the `Exploring Generative AI curriculum` not use OpenAI?  Is that why they can be safely left as `...`?  If so, will close this comment out as that should be fine for gemini too.

sanchitmalhotra126 commented on config/development.yml.erb:
```diff
@@ -70,6 +70,8 @@ openai_evaluate_rubric_api_key:
 openai_student_learning_api_key: !Secret
 openai_measures_of_learning_api_key: !Secret
 
+google_gemini_student_learning_api_key: !Secret
```
nit: I don't know if this is strictly necessary, as `config.yml.erb` serves as the default for the various `<environment>.yml.erb` configuration files. I believe we typically only override here if we want a different environment-specific value (@bencodeorg correct if I'm wrong though? looks like we do have other openai-* values in both config and development.yml.erb)

edcodedotorg replied:
yes I matched openai, so let me know if I should remove both gemini and openai settings.  I do vaguely remember thinking exactly what you wrote and that maybe it didn't work when removing though

edcodedotorg replied:
I just tried again by removing` google_gemini_student_learning_api_key: !Secret` from `config/development.yml.erb` and it looks like it is needed there. Closing this out as the current settings mimic openai so if there is an improvement, it can be done for both

<img width="932" alt="Screenshot 2025-06-26 at 2 47 40 PM" src="https://github.com/user-attachments/assets/6a2518b0-d474-420b-b9ce-b0d73a5599d6" />

sanchitmalhotra126 commented on dashboard/app/helpers/aichat_ai_client.rb:
```diff
@@ -0,0 +1,274 @@
+    level = Level.find_by(id: level_id)
+
+    # level system prompt - string or nil
+    # TODO - determine if we're still using level system prompts
```
Let's update this TODO to something like "pass in level system prompt as a separate argument" or "retrieve level system prompt in a level type-agnostic way" as currently this assumes the level type is AI Chat and therefore has an `aichat_settings` field (not true for AI Tutor v2 levels and beyond).

sanchitmalhotra126 commented on dashboard/app/helpers/aichat_ai_client.rb:
```diff
@@ -0,0 +1,274 @@
+      end
+    end
+
+    # Note for PR (TODO - discuss and remove this note before PR merged):
```
We can keep as is; as this logging is read and queried via Cloudwatch logs, we tend to prefer camelCase.

sanchitmalhotra126 commented on dashboard/app/helpers/aichat_gemini_client.rb:
```diff
@@ -0,0 +1,135 @@
+
+  # The url to send with the post request
+  private def url
+    # TODO secret will be per product (ai chat vs tutor) - currently we have just one for both
```
nit: maybe *key or *api_key rather than "secret" (for reader clarity)

edcodedotorg replied:
sure, good point - I had to remember why I even called it that.  Secret must have been referring to AWS Secrets Manager which is where it's stored

edcodedotorg replied:
comment removed and put in general TODO

sanchitmalhotra126 commented on dashboard/app/helpers/aichat_gemini_client.rb:
```diff
@@ -0,0 +1,135 @@
+    # --------------------------------------------------------------------
+    # We should probbaly expand the usage metrics to include thinking metrics
+    # but will need to deal with cases where given backend do not support them
+    # --------------------------------------------------------------------
```
This sounds like something we can file a follow-up ticket for and revisit.

breville commented on apps/src/lab2/ai/AiTutor2Manager.ts:
```diff
@@ -18,19 +19,28 @@ const systemPrompts = {
 export type AiTutor2MessageType = 'hint' | 'user';
 
+// This type is the union of all the valid AI Model IDs
+// Note: this type seems out of place, based on name `AiChat` but we will eventually
+//  need to  come up with a name for the shared plumbing (for AI Chat and AI Tutor)
+//  and then this will take on that name and can be move accordingly
```
nit: this feels like a TODO which can be tracked elsewhere.  I think it's best for our comments to capture the current state of affairs concisely.

edcodedotorg replied:
yup, we're moving all these to the PR. Thanks for flagging!

breville commented on apps/src/lab2/views/components/aiTutor2/useAiTutor2.tsx:
```diff
@@ -22,8 +25,20 @@ export function useAiTutor2(
+  const modelQueryParam = queryParams('aitutor2-model');
+
+  // Only use modelQueryParam as modelId if it is a valid id, otherwise set to undefined
+  //  which lets AiTutor2Manager use its default
```
nit: no need for an indent on the second line.  Also, we should end all comments with a period.

edcodedotorg replied:
ok I'll update this one, not sure I'll catch all other indents today as I'm in the habit of doing that (I find it more readable)

breville commented on dashboard/app/helpers/aichat_ai_client.rb:
```diff
@@ -0,0 +1,252 @@
+    raise_not_implemented_error
+  end
+
+  # take response_body and raise any errors if appropriate
```
nit: All comments should start with an uppercase letter and end with a period.

edcodedotorg replied:
ok will leave this open to look through all files

edcodedotorg replied:
Just pushed a commit that follows this rule for all files I had open (so some that I otherwise didn't touch).  Closing this as completed

breville commented on dashboard/app/helpers/aichat_ai_client.rb:
```diff
@@ -0,0 +1,252 @@
+  end
+
+  # Note - this is temporarily disabled for gemini by overriding this method there with a "no-op"
+  # TODO - provide mechanism for metric name to be based on provider.  This will require updating
+  #  queries and dashboards as explained here:
+  #  https://github.com/code-dot-org/code-dot-org/pull/66675#discussion_r2164695066
```
Can we track this TODO elsewhere?

edcodedotorg replied:
yup all todos are being removed from comments to be tracked in this PR.  Missed that one thanks

breville commented on dashboard/app/helpers/aichat_ai_client.rb:
```diff
@@ -0,0 +1,252 @@
+      count: prompt_tokens
+    )
+
+    #calculate costs
+    input_rate = 0.15 / 1_000_000 # $0.15 per million tokens
+    cached_input_rate = 0.075 / 1_000_000 # $0.075 per million tokens
+    output_rate = 0.60 / 1_000_000 # $0.60 per million tokens
```
Can we calculate financial information outside of public code?

edcodedotorg replied:
not sure how we'd do that unless this logging goes or would go through a lambda or something (don't know much about it)?  But this is just copy/pasted from existing code.  Should I add this to the general TODO comment?

bencodeorg replied:
Yeah not sure I see the issue here Brendan? This just reflects public OpenAI pricing information for gpt-4o-mini.

edcodedotorg replied:
discussing offline, waiting for additional guidance

edcodedotorg replied:
moved to comment in PR

bencodeorg commented on dashboard/app/helpers/aichat_ai_client.rb:
```diff
@@ -0,0 +1,252 @@
+# methods. Currently the two implemented APIs (OpenAI and Gemini) are POST based REST APIs
+class AichatAiClient
+  # create an instance of the appropriate derived class based on model id
+  def self.create_instance(model_id)
```
This is a bit confusing to me, is this standard? That the parent class knows about exactly which subclasses exist?

bencodeorg replied:
I saw mention of the "factory" pattern earlier, maybe that's what this is?

edcodedotorg replied:
yeah pretty common: https://en.wikipedia.org/wiki/Factory_method_pattern

edcodedotorg replied:
yeah if you think about the related classes as a family,  it kind of makes sense that the base class (parent) knows about its derived classes' (children's) capabilities. The key here is separation of concerns (if not what we normally think of as encapsulation) as it's better for the base class to know about which of the family members is best able to do the job then to let this leak out (where the caller/instantiating-code) would otherwise need to be aware of something that is best kept in the family (hope that makes sense - been staring at code too long)

bencodeorg replied:
That makes sense to me!

breville replied:
Worth noting that in a classic implementation, the factory method could be separate to the base class, which leaves the base class (often nothing more than a pure abstract interface) not knowing anything about the concrete subclasses.  The value of the pattern is that other code in the system only has to use that base interface.

edcodedotorg replied:
@breville. Sure, from the wikipedia I shared above (emphasis mine):

> Factory methods can be specified in an [interface](https://en.wikipedia.org/wiki/Interface_(object-oriented_programming)) and implemented by subclasses **_or implemented in a base class_** and optionally [overridden](https://en.wikipedia.org/wiki/Method_overriding) by subclasses

So I think this is a common enough pattern that there are actually many common variations but if there is an established practice within the org, I'm happy to follow it.

With that said, Ruby doesn't have interfaces and doesn't have abstract classes or methods but I simulated abstract methods here by raising a not implemented exception.  If this was say TypeScript (or Java or C#), that supports pure interfaces and abstract classes with concrete methods (unlike say C++ that you have to simulate interfaces with abstract classes - haven't thought about that in a while), I personally would have implemented this in essentially the same way. 

I would have used an interface to define the shape of (at least) the derived classes and the base class would be there  to provide some common shared functionality with its private concrete methods (`initialize`, `file_is_image?`, `file_is_pdf?`, `get_messages_with_assets_count`, etc - whether the latter ones actually make sense as methods vs. functions is also a style question in many languages).  

From the outside, we have one public method (`get_response_text`) so rather than create a completely separate class to hold the factory method (or alternatively a standalone function) I just added one more public method to the base class (so now it knows how to `create_instance` and `get_response_text`).  This might seem like a style choice but what it allows is for the decision of which subclass to use when creating an instance, to be decided or influenced by private methods of the subclass themselves, which often becomes useful (the patterns I follow or create are often based on my experience wrt how the code might evolve, not where it is now).  Of course this can be accomplished by making the factory a "friend" of the class family for languages that support that (TypeScript doesn't and neither does Ruby IIRC) but in the case of a base class that already has a single public concrete method (`get_response_text`) I actually find it appropriate to give it a second, which just happens to be a factory method.  

Hope that makes sense, but again if there is an established practice within the org/codebase (or your experience) that actually covers all thes nuances here, I'm happy to learn and adapt.

breville replied:
Oh that's fine, my comment was mainly for Ben to give a little more context about the use of factories.

bencodeorg commented on dashboard/app/helpers/aichat_ai_client.rb:
```diff
@@ -0,0 +1,252 @@
+  # private initializer - all instances should be created with `create_instance` factory
+  private def initialize(api_key, model)
+    @api_key = api_key
+    @model = model
```
Am I reading right that this is the same as what we call `model_id` everywhere else? I think it'd be clearer to make them match if they are the same?

edcodedotorg replied:
good eye, but no.  This is what I keep talking about regarding a hash being needed to go between our internal model_id and the model we send the the 3rd party API.  Notice the initializer is private so it should only be created through this `create_instance` call.  And notice how only in the case of gemini is the model (as sent to 3rd party) and (internal) model id the same:

<img width="948" alt="Screenshot 2025-06-26 at 5 48 16 PM" src="https://github.com/user-attachments/assets/4ab57a5c-aa77-451e-93b8-4876d45bf524" />

bencodeorg replied:
Gotcha, makes sense they're different then. Thanks!

edcodedotorg replied:
yeah hopefully once we tackle the hash it'll be clearer

edcodedotorg commented:
Hi All,  Thanks for bearing with my various TODOs and notes as comments in the code...

That was the only way I was going to be able to keep track of all those open questions and potential improvements as I was working on this refactoring.  

I've consolidated them all here into this comment but as some of the context is lost, it might be necessary to go back to the code, where these comments were inline, to make sense of some of them.  Here is the [commit right before I started pulling them out of the code](https://github.com/code-dot-org/code-dot-org/tree/3c02d711ca6672a470fdd64a11eb42bd8fa5853b) in case that's useful.

## TODOs
### Short term 
1. add `report_usage_and_throttling_metrics()` call back to gemini class. This requires some thought around how to make it implementation agnostic re: 'messages' and 'usage’. This will require updating queries and dashboards as explained in Sachit’s comment.  Metrics should also be generalized as the current implementation of `get_usage_from_body` is based on OpenAI’s chat completions API
2. Derive an `aichat_sagemaker_client.rb` from ` aichat_ai_client.rb` to type Sagemaker into new class structure
3. Kill `aichat_ai_helper.rb` entirely.  Everything in there could be either pushed up to active job or down into the shared class (or combination of the two).  This would encompass moving token_throttling_key to new class
4. We need a mechanism to support multiple keys per backend as gemini key and open AI keys will be separated by product (AI Chat/AI Tutor) for billing purposes . The idea is to create a hash from model_id => product_id (where model_id is internal model id and product_id is AI Chat/AI Tutor/etc) =>  model/key/endpoint (all of which are to determine how/which API to talk to)
5. Add “filename” support for gemini which is necessary for a user to reference a given file when multiple are uploaded. This is not possible natively in gemini but can be handled by concatenating a message that lists the attached files.  This will need some thought as the concatenated message will need to be in the language based on the user's locale (e.g. Spanish)
6. Move to OpenAI’s responses API endpoint


### Medium Term
1. Improve file type detection. `file_is_image?` currently uses the previous logic where if not a pdf, it's an image, which is a bit flaky.  We should probably tie it to the allowed image extensions. Ideally, the file type should be handled with mime-type as file extension does not determine file type, but the way “attach file“ works it seems to always append a specific file extension to the file before uploading, if that file extension is not present
2. Raise just our custom exception classes from `AichatAssetHelpe.rb` (by rescuing e.g. AWS errors)   and get a descriptive message back to the user explaining that it was an internal error but related to the file attachment.
3. Pass in level system prompt as a separate argument" or "retrieve level system prompt in a level type-agnostic way" as currently this assumes the level type is AI Chat and therefore has an aichat_settings field (not true for AI Tutor v2 levels and beyond).
4. Improve `ai_client_test` tests.  
    1. While we currently only test a single hardcoded model id for gemini (gemini-2.0-flash) we should modularize so that the model id is dynamically added to the body (once we have hash of model_id => model/key/provider) 
    2. test that key is passed as query param for gemini
    3. use the hash to go from internal_model_id to model_id (since these aren't same for openai) and switch based on different models
    4. look for other general areas of improvement
5. Look into whether there is any benefit for files to come first in generated content list in `format_content_item`. This seems more logical regarding how it is currently displayed in the UI (above text) and thus possibly how user may refer to it (e.g. “please reference the previous files and analyze xyz”)
6. Figure out a name for the shared plumbing (for AI Chat + Tutor + beyond).  
    1. Once we have that name many things can be renamed and or moved e.g. `export type AiChatModelIdType = ValueOf<typeof AiChatModelIds>`
    
### Long term
1. Add multimodal response support to `aichat_ai_client.rb` and rename `extract_text_response_from_body`.  My preference is to do this before we have any use case as it will enable exploration and it will force us to find the right method “shape” (params and return value) to support the various types of responses before we get too far with this implementation 
2.  Expand the usage metrics to include “thinking” metrics. To do this we will need to allow for a given backend that does not support them. This could be generalized as “optional” metrics

bencodeorg commented on apps/src/lab2/ai/AiTutorModelId.ts:
```diff
@@ -0,0 +1,17 @@
+const modelQueryParam = queryParams('aitutor2-model');
+
+// This type is the union of all the valid AI Model IDs.
+export type AiChatModelIdType = ValueOf<typeof AiChatModelIds>;
```
This is defined in `AiTutor2Manager.ts` as well? I think they're the same thing, yeah?

edcodedotorg replied:
good catch, updated to import from there.  thanks

bencodeorg commented on dashboard/test/helpers/aichat_gemini_client_test.rb:
```diff
@@ -69,63 +69,58 @@ class AichatGeminiClientTest < AichatAiClientTest
-        assert_equal subject, @response_text
+      context 'when body is well formed and request fails with error JSON' do
+        let(:new_message) {@new_message}
+        # Don't expect any particular fields in body, we're just testing that we
```
"testing that we..."?

edcodedotorg replied:
fixed, thanks

bencodeorg commented on dashboard/test/helpers/aichat_gemini_client_test.rb:
```diff
@@ -69,63 +69,58 @@ class AichatGeminiClientTest < AichatAiClientTest
+
+        let(:stubbed_response_body) {stubbed_success_response_body}
+        it 'successfully makes request and is returned the correct response' do
+          # Check that we're returned the correct response.
```
Typo: we've?

edcodedotorg replied:
fixed, thanks

bencodeorg reviewed (commented):
Just looked at the two commits from today but look reasonable to me in general!
